File: divide-and-conquer/construct-quad-tree/ConstructQuadTree.java
/*
//Question: https://leetcode.com/problems/construct-quad-tree/?envType=study-plan-v2&envId=top-interview-150

// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
}
*/

class Solution {
    public Node construct(int[][] grid) {
        int size = grid.length;
        return build(grid, 0, 0, size);
    }

    private Node build(int[][] grid, int rowStart, int colStart, int size) {

        if(size == 1) {
            boolean val = false;
            if(grid[rowStart][colStart] == 1) {
                val = true;
            }

            return new Node(val, true);
        }

        int half = size/2;
        Node topLeft = build(grid, rowStart, colStart, half);
        Node topRight = build(grid, rowStart, colStart+half, half);
        Node bottomLeft = build(grid, rowStart+half, colStart, half);
        Node bottomRight = build(grid, rowStart+half, colStart+half, half);

        
        if(topLeft.isLeaf && topRight.isLeaf && bottomLeft.isLeaf && bottomRight.isLeaf 
        && topLeft.val == topRight.val && topLeft.val == bottomLeft.val && topLeft.val == bottomRight.val) {
            return new Node(topLeft.val, true);
        }

        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }
}

File: divide-and-conquer/construct-quad-tree/construct_quad_tree.py
"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution:

    def construct_quad_tree(self, grid: List[List[int]], row_start: int, column_start: int, size):

        if size == 1:
            val = False
            if grid[row_start][column_start]==1:
                val = True
                
            return Node(val, True, None, None, None, None)

        half = int(size/2)
        topLeft = self.construct_quad_tree(grid, row_start, column_start, half)
        topRight = self.construct_quad_tree(grid, row_start, column_start+half, half)
        bottomLeft = self.construct_quad_tree(grid, row_start+half, column_start, half)
        bottomRight = self.construct_quad_tree(grid, row_start+half, column_start+half, half)
        
        if (
            topLeft.isLeaf and topRight.isLeaf and
            bottomLeft.isLeaf and bottomRight.isLeaf and
            topLeft.val == topRight.val == bottomLeft.val == bottomRight.val
        ):
            return Node(topLeft.val, True, None, None, None, None)

        return Node(True, False, topLeft, topRight, bottomLeft, bottomRight)

    def construct(self, grid: List[List[int]]) -> 'Node':

        row_start = 0
        column_start = 0
        return self.construct_quad_tree(grid, row_start, column_start, len(grid))